Added support for mouse dragging

NormySan opened the pull request:
Hey,

I've added support for mouse dragging including some tests and an example that has clickable links that can be clicked while not dragging and will prevent any clicks while dragging.

I also updated the typings and the readme.

Did some minor refactoring on the Slider methods so that both mouse dragging and touch dragging use the same logic for setting state and the animations.

Please let me know if you want me to change anything!

Best Regards
Christoffer

mrbinky3000 commented:
Thanks a lot!  Looking over this now.

mrbinky3000 commented on src/Slider/Slider.jsx:
```diff
@@ -88,28 +97,102 @@ const Slider = class Slider extends React.Component {
+  handleOnMouseUp() {
+    if (!this.props.dragEnabled || !this.state.isBeingMouseDragged) return;
+
+    setTimeout(() => {
```
Not a big fan of setTimeout in code.  Can you describe the problem you encountered that required the use of a timer?

NormySan replied:
I totally understand, the reason i had to add it was to be able to handle clicking on slides that may contain a link but at the same time also not having clicks go through when you're dragging.

So to not have the state change immediately when you release the mouse button and set isBeingMouseDragged & mouseIsMoving to false which would let a click go through i had to delay it until the click has been registered.

TheFullResolution replied:
What about using `requestAnimationFrame`?

I am working on a component which allows dynamic content inside carousel, I used the method described in below blog post to calculate naturalWidth and naturalHeight, works great without magic setTimeouts.

https://stanko.github.io/react-rerender-in-component-did-mount/

mrbinky3000 replied:
Any timer combined with setState is a bad idea.  The reason being that there's a chance that react might re-render and setState will reference an instance of the class that no longer exists.  You end up with a lot of warnings in the console.  handleOnMouseUp sets the timeout which calls onDragEnd which contains a setState.  Another thing about timers is that 100 might not be enough time if a user's computer is lagging for some reason.

mrbinky3000 replied:
Let me think about this one for a bit.  I think there may be some css way to prevent the click event from bubbling.

NormySan replied:
Yeah, it would definitely be best if we can have a solution without a timeout if possible.

NormySan commented:
So i removed the on mouse up event handler and moved the logic in that handler to the click handler since a click is always registered when the mouse is released this has the same effect and lets us remove the timeout.

I played around a bit with the tests but I'm not able to get them to 100% and I'm not really sure how to fix it. It's saying that line 167 is uncovered.

tim-steele commented:
@NormySan Is this the line in question?

```
if (this.state.isBeingMouseDragged && this.state.mouseIsMoving) {
    ev.preventDefault();
}
```

NormySan commented:
@tim-steele Yes

tim-steele commented:
@NormySan when you use `.simulate()` it takes a second param, and that is the data being passed to the event handler.

So, to test `preventDefault()` being called, you would want to set up a Jest function and the component:
```
const mockPreventDefault = jest.fn();
const mockedComponent = shallow(<ComponentHere />);
```

Then set the state of your component:
```
mockedComponent.setState({ isBeingMouseDragged: true, mouseIsMoving: true });
```

and then do a simulate on the object that triggers the handler, like so:
```
object.simulate('mousedrag', { preventDefault: mockPreventDefault });
```

Finally, the test you would run/expect would be the preventDefault being called:
```
expect(mockPreventDefault).toHaveBeenCalled();
```

Let me know if you have any questions about this, but hopefully this helps.

NormySan commented:
Thanks, I'll see if i can get the tests done next week.

tim-steele commented:
@NormySan I was taking at look at line 167 to see if I can help you (so your PR doesn't get stale). I was looking at your code, and it looks like the reason it can't be tested is because it is "unreachable".

```
  handleOnMouseClick(ev) {
    if (!this.props.dragEnabled || !this.state.isBeingMouseDragged) return;

    if (this.state.isBeingMouseDragged && this.state.mouseIsMoving) {
      ev.preventDefault();
    }

    this.onDragEnd();
  }
```

This states that if either `dragEnabled` or `isBeingMouseDragged` is *not* `true`, then return/exit function. 

Then it tests if _both_ `isBeingMouseDragged` and `dragEnabled` is `true`, then fire `preventDefault()`. However, if they weren't both true, it would have already exited the function. It looks like an unnecessary `if` statement. The only way to be at that point in the code is if both `isBeingMouseDragged` and `dragEnabled`are `true`. 

`preventDefault()` is being tested correctly in your PR. It is the untested else statement that is failing. If you remove the `if` your tests should pass and it shouldn't break anything.

NormySan commented:
Thanks, I see what you mean. I've been sick so i haven't had time to look at this, hopefully i can get it done this week :)

NormySan commented:
Hum, I'm still not really sure i get it. I see that the check on `this.state.isBeingMouseDragged` can be removed but the second if check with `this.state.mouseIsMoving` is required otherwise any links contained in the slider can never be clicked.

The second if check is there to only prevent clicks when the slider is in motion, if it's not a click on it should work like it does normally e.g. it should not be prevented.

This works perfectly fine when trying it out in the browser and it's most definitely reachable code but when testing it with Jest and simulating the exact same events it's for some reason considering line 167 as uncovered even tough the preventDefault has been reached and an assertion run on it.

It may be that i totally misunderstand how coverage works, if so then it would be great if someone could help me out on how to cover line 167.

If the code needs to be changed somehow it needs to be a change that allows clicks to go through when the slider is not being dragged but prevented if it is being dragged.

tim-steele commented:
I think the problem then is the first `if` statement.

```
handleOnMouseClick(ev) {
    if (!this.props.dragEnabled || !this.state.isBeingMouseDragged) return;

    if (this.state.isBeingMouseDragged && this.state.mouseIsMoving) {
      ev.preventDefault();
    }

    this.onDragEnd();
  }
```

This line `if (!this.props.dragEnabled || !this.state.isBeingMouseDragged) return;`.

If `this.props.dragEnabled` **or** `this.state.isBeingMouseDragged` is `false` then your function immediately exists/returns. The only way you can go to the next line is if **both** `this.props.dragEnabled` and `this.state.isBeingMouseDragged` is `true`.

So essentially `if (this.state.isBeingMouseDragged && this.state.mouseIsMoving)` and `if (!this.props.dragEnabled || !this.state.isBeingMouseDragged)` are the same exact test. That is why it can't test the else, it would never get to line 167 if `this.props.dragEnabled` or `this.state.isBeingMouseDragged` is `false`.

NormySan commented:
But the test is checking that the `ev.preventDefault();` has been called and this only happens if the code reaches the second if statement and if it's true.

But maybe Jest isn't smart enough to know this? Maybe i need additional tests just to trigger the actual function with specific state set and not just simulating mouse events?

tim-steele commented:
It **is** testing preventDefault(). The failing part of the test is the branching. It is saying it can't reach 100% code coverage because Jest can't test the `else` condition of the second `if` statement. That is because the two `if` statements are saying the same thing.

![image](https://user-images.githubusercontent.com/1490225/37785457-926c8020-2dd0-11e8-81fc-da7f44a947e6.png)

tim-steele commented:
So you are failing the 100% test coverage on [branching](https://stackoverflow.com/questions/35034977/what-is-a-branch-in-code-coverage-for-javascript-unit-testing#answer-35035591).

Every time there is an if statement it tests if it resolves to `true` and resolves to `false`. In this situation, the first `if` statement prevents this second `if` statement from ever resolving to false. So the second `if` statement is redundant.

In re to `this.state.mouseIsMoving` being needed: if `this.state.mouseIsMoving` is ever `false` the first `if` statement returns. 

That is why it is a redundant `if` statement. The second `if` statement will _always_ resolve to `true` the way written, and so you can never test the other branch (`else`).

NormySan commented:
But it will, if the mouse button has been pressed and dragging is enabled the the first if statment will pass and then the second if statment will check if the mouse is in a moving state e.g. dragging the carousel.

If the user has not started dragging the click will not be prevented when the mouse is released but just continue on and do what it must to end the dragging.

The check for `this.state.isBeingMouseDragged` can be removed in the second if statment but `this.state.mouseIsMoving` must be there since it can be true or false depending on if the mouse has been moved or not while in a dragging state.

tim-steele commented:
I am very sorry. I made a chart, and I realized I was reading the props wrong (`dragEnabled` vs. `mouseIsMoving`). I see what you are saying. Again, I am _really_ sorry about that. 

I took a look at your "should not prevent default action on clicks when not dragging or mouse moving" test, and if you make the following adjustment you get 100% code coverage.

```
  it('should not prevent default action on clicks when not dragging or mouse moving', () => {
    const wrapper = shallow(<Slider {...props} dragEnabled />);
    const instance = wrapper.instance();

    instance.sliderTrayElement = {
      clientWidth: 100,
      clientHeight: 100,
    };

    drag100.preventDefault.mockReset();

    wrapper.setState({
      isBeingMouseDragged: true,
      mouseIsMoving: false,
    });

    wrapper.find('.sliderTray').simulate('click', drag100);
    wrapper.update();

    expect(drag100.preventDefault).toHaveBeenCalledTimes(0);
  });
```

NormySan commented:
Thanks, I'll fix the test and update the pull request sometime this week :)

mrbinky3000 commented:
How goes this?  Got another PR in the works.  Trying to minimize collisions.

NormySan commented:
Sorry, I've had a lot to do, thanks for notifying me, just committed updated tests with full coverage.
